### ISSUE TITLE
Memory leak in stats

### ISSUE THREAD

**Issue by BlueSilver22** [BTCPrivate/z-nomp#3]
There is a memory leak somewhere in the stats of z-nomp. We are offering a 10 BTCP bounty to locate and fix this issue.

**Comment by rewrewby** [comment]
Check this:
https://github.com/BTCPrivate/z-nomp/pull/4

**Comment by Patrickdurbin** [comment]
https://github.com/BTCPrivate/z-nomp/blob/master/libs/stats.js

The issue is here  its reflecting back into api/stats

issue was carried over

https://gitter.im/zclassicorg/z-nomp?at=5a8857d935dd17022eb7f84b

**Comment by rewrewby** [comment]
stats.js is for json requests like /api/stats.
/api/live_stats making heavy load on server.

**Comment by Patrickdurbin** [comment]
https://github.com/BTCPrivate/z-nomp/blob/master/libs/stats.js
The issue is here its reflecting back into api/stats
issue was carried over
https://gitter.im/zclassicorg/z-nomp?at=5a8857d935dd17022eb7f84b

**Comment by davidtai** [comment]
The problem that its adding like 35mb in payout data, I think you disable that and its okay.  Let me whip up a quick PR.

**Comment by egyptianbman** [comment]
IMHO, The best solution is to break down these few but **massive** api calls (they return everything under the sun) into more, but smaller api calls that return specific data. This will require more changes than a quick fix but it will allow for much more efficient bandwidth usage and will be easier on the servers.

**Comment by sennevb** [comment]
remove live_stats indeed...

**Comment by davidtai** [comment]
That won't fix the problem.  The huge amount of payouts is generated from the redis call in getGlobalStats.  The static template rendering calls it:

https://github.com/z-classic/z-nomp/blob/ff0e1f6f32150bf56b2e3f35530a53714c29f064/libs/website.js

This is also why when you go to the stats page the bottom 2 sections don't load.

@BlueSilver22, I put out a quick PR for you. #5

**Comment by davidtai** [comment]
Just FYI, its not a 37mb request.  The event source connection gets to 37mb and explodes.  Sometimes if its feeling lucky it will try to restart the connection.

Each request is actually about 5mb.  That's 5mb for every time a page is hit and every few seconds a person is connected to a stats page.  

I'd recommend disabling the entire payout query as a next step.  Then we'll need to look at the workers hash.  Here's a copy of a request for reference.

[lotsofpayouts.txt](https://github.com/BTCPrivate/z-nomp/files/1781899/lotsofpayouts.txt)

**Comment by coinblockers** [comment]
caching is king

**Comment by BlueSilver22** [comment]
@coinblockers we already cache quite a bit. But this does not prevent the memory leak crash

**Comment by ConTrast77** [comment]
I apologize if I'm weak in programming and distract you from work.
libs / api.js

Line 46, 47 - accumulated objects !!!, each time reset! (initialization)?
 var history = {};
 var workers = {};
Line 65 - zeroing as an array!
history [w] = [];

Line 68 - Alternative - array accumulation and its cells "w" = array.Object/values​​ (and possibly rashly eliminates/skips the condition of line 65)
history [w] .push (...)

lines 79, 82, 83, 87, 88 - cumulative array.object
   workers [w]. ...

---
At the same time, there is no output data (they are not used), they are constantly growing, but there are other dependent alternative objects, like
line 88
totalHash + = portalStats.stats.pools[pool].workers[w].hashrate;

--
workers[w]. ...
and
portalStats.stats.pools [pool].workers[w]

not the same, connections i do not see or do not understand.

**Comment by ConTrast77** [comment]
libs / api.js
All switch() {case }
-ended of "return;"
- only "case 'getblocksstats': break; " at line 32
And the height (ladder "case") of application - affects the breaking.

**Comment by oceanchild** [comment]
@ConTrast77  those local variables are garbage collected.

**Comment by ConTrast77** [comment]
If the hardware protection does not lead to a hangup. Could it be that the hardware error, not the programming? Is there any reporting on this in server protocols? ))

+ What are the limitations of Java / Brouser?

**Comment by antisocialmunky** [comment]
@BlueSilver22 can you give us a branch with your current modifications?  I'm having quite a difficult time reasoning about the code while not knowing if what's in master is up to date.

**Comment by nimbosa** [comment]
yes, just to clarify as @antisocialmunky pointed out, @BlueSilver22 :

is this branch of Z-nomp uo to date and what is used LIVE on the live BTCP explorer website?

**Comment by BlueSilver22** [comment]
@interbiznw and @michaelotis are best to ask for the branch we are working off of.

As of right now, we are thinking of moving to MPOS as z-nomp seems poorly equipped for scalability
